core/src/leader_schedule.rs

```diff
@@ -73,4 +73,28 @@ mod tests {
         // Check that the same schedule is reproducibly generated
         assert_eq!(leader_schedule, leader_schedule2);
     }
+
+    #[test]
+    fn test_repeated_leader_schedule() {
```

Comment:
How about a more minimal test? This test completely repeats the implementation. It's trivially true. How about something that aims to show how setting `repeat` results in something different than leaving it set to one? Maybe something like:

```rust
let alice_pubkey = Keypair::new().pubkey();
let bob_pubkey = Keypair::new().pubkey();
let stakes = vec![(alice_pubkey, 2), (bob_pubkey, 1)];

// What the schedule looks like without any repeats
let leaders1 = LeaderSchedule::new(&stakes, seed_bytes, len, 1).slot_leaders;
let leaders2 = LeaderSchedule::new(&stakes, seed_bytes, len, 2).slot_leaders;
assert_eq!(leaders1.len(), leaders2.len());
assert_eq!(leaders1, leaders1_expected);
assert_eq!(leaders2, leaders2_expected);
```

Reply:
@garious , given the randomness of the generated schedule, it will be tough to compute leaders1_expected, and leaders2_expected. I tried writing a similar test. In leaders1 schedule came out to be [alice, bob, alice, alice]. Not sure if that's deterministic and can be pre-filled in the expected result.

Reply:
Yes, it's deterministic, but you can't use: `let seed = Keypair::new().pubkey();` . Try `let seed = Pubkey::default();`

Reply:
@garious , it's coming out to be this. Is this what you are expecting? I had to run the test to figure out what expected values should be, and then update the test to make it pass.

```
        let alice_pubkey = Keypair::new().pubkey();
        let bob_pubkey = Keypair::new().pubkey();
        let stakes = vec![(alice_pubkey, 2), (bob_pubkey, 1)];

        let seed = Pubkey::default();
        let mut seed_bytes = [0u8; 32];
        seed_bytes.copy_from_slice(seed.as_ref());
        let len = 8;
        // What the schedule looks like without any repeats
        let leaders1 = LeaderSchedule::new(&stakes, seed_bytes, len, 1).slot_leaders;

        // What the schedule looks like with repeats
        let leaders2 = LeaderSchedule::new(&stakes, seed_bytes, len, 2).slot_leaders;
        assert_eq!(leaders1.len(), leaders2.len());

        let leaders1_expected = vec![
            alice_pubkey,
            alice_pubkey,
            alice_pubkey,
            bob_pubkey,
            alice_pubkey,
            alice_pubkey,
            alice_pubkey,
            alice_pubkey,
        ];
        let leaders2_expected = vec![
            alice_pubkey,
            alice_pubkey,
            alice_pubkey,
            alice_pubkey,
            alice_pubkey,
            alice_pubkey,
            bob_pubkey,
            bob_pubkey,
        ];

        assert_eq!(leaders1, leaders1_expected);
        assert_eq!(leaders2, leaders2_expected);
```

Reply:
Yes, something like that is what I was expecting. Then it's easy to see that `repeat` can push all other stakers right off the schedule. The result isn't a fair distribution.  Instead of repeat, how about we group all the duplicates without otherwise reordering them?

Reply:
So in that case, every node will still get same number of leader slots as it would've got without this change. It's just that the slots will be consecutive. Is that right?

May be the number of consecutive slots can have an upper-bound, unless there's no other node scheduled to be a leader in that epoch...

Reply:
I agree, but before you implement that upperbound, let's think about the details of that attack and write it down.  We might want a lower bound as well, but for performance reasons. For now, let's keep the code simple and readable.

Reply:
So, no upper/lower bound for now. Just rearrange the leader schedule so that leader slots for a node are consecutive..

Reply:
with really long epochs (i.e. slots per epoch >>> consecutive_slots), we don't need to worry about unfairness...

Reply:
the goal of this change is to reduce the percentage of network downtime that results from grace periods.  keep voting times short while leaders stay leaders long enough to wash out network latency